### ISSUE TITLE
"higher priority first" or "closed distance preferred"

### ISSUE THREAD

**Issue by hellas2000** [TheDogKSP/MoreEffectiveTransfer#22]
Hi, I'm a big fan of your MOD, thank you for your hard work. This MOD should improve the game experience with no doubt.

I have use the MOD on two of my cities. One is a small city in a small area with 60k people, which works good so far. 

The other is a larger city, and a special one. It has 6 towns separated with long distances, and connected by highways. In this city the MOD seems not working well, so many emergency icons sparkle everywhere. Because some districts are so small for some special purposes, I didn't select the option "Prefer local district services". In order to test this problem, I have put a few dozen extra service buildings, it definitely should be enough, but the problem still exists. 

I know this is a very much tricky problem, as you said in the MOD description 'There is obviously a conflict between the "higher priority first" approach and the "closed distance preferred" approach."'. I have an interest in finding out if this MOD cound solve the problem, I'm wondering if you could add an option for user to choose "higher priority first" or "closed distance preferred", if this is not too much work. And if user choose "closed distance preferred", then priority is concerned only when the distances are "almost" the same. And I would like to test it on my city.

thanks.

**Comment by TheDogKSP** [comment]
Thanks for the feedback.

Indeed it is very tricky, and over the past versions there have been some changes in the match-making logic to alleviate some issues that were observed, most notably the death care issue (hearses going inefficient routes so dead body problems are not resolved quickly enough).

Funny enough, the "intuitive" approach of always preferring the closest service led to the worst performance compared to vanilla (if looking at death care or garbage).

The current logic is much closer to vanilla, but of course exhibits more vanilla-like behaviour, such as vehicles going longer routes that you would expect.

What you are asking for is definitely possible, just quite some additional development effort (quite some more that a simple "if (option) then this, else that..."). 
But I am very willing to consider this and try it out with you. Maybe first not as an option in the general mod, and more of a custom version compiled for you to test, if you dont mind.
Feel free to add me on steam contacts to further discuss this.

Cheers, cambrium/thedog

**Comment by TheDogKSP** [comment]
Also, for your information, adding some insights into what is going on "under the hood":

Service Vehicles being "pulled out of position" across the map is a problem that is always happening, be it vanilla, metm oder metmm continued (with or without prefer local option), due to the fact that usually DEMAND is a lot higher than SUPPLY of a service.

As match-making is being performed at a specific point in time, all open demand and supply for a service are matched.
So it is quite possible, even likely, that at one point in time the following happens:
- most service vehicles are in use and busy
- lots of buildings demand service
=> for a given building the "closest" service vehicle might still be across the whole map, just because every closer service vehicle is already in use.

Usually the effect is, that say a police car comes rushing over the highway, although the target building has a police station around the corner. 
The match-making, however, has zero knowledge about this. All it sees are the open demands and supplies at a point in time.
Even less does the match-making know that if the target building were patient and waited a bit longer, a much closer vehicle would be available 15 seconds later. :(

All this is not meant as an excuse for how METM continued performs. It is just the reality how the transfer manager system in CSL works.
And I can guarantee that you will always see this effect, regardless of how the algorithm of match-making is changed to prefer closer distances etc.


There is basically only one full solution to the problem:
Hard-limiting every service to specific districts. This is what EDS did, with lot of micro-management required.
I opted not to implement it, and chose to go with a softer limit system ("prefer local service"). But the longer the game runs, the higher the chance that a vehicle WILL be piulled out of position by a request that just does not have any closer vehicle nearby (at that specific point in time).

Anyway, lets try how algorithm changes affect your specific scenario :)

**Comment by hellas2000** [comment]
Hi TheDogKSP, thank you for your reply,

I would love to try your custom version, It's totally ok.  And I hope my test result will do some help. I will add you on steam contact, if you have anything for me to test, let me know.

Limit service vehicles to districts maybe a solution, but the district has already been used in several situations, I think it will only work if there is a new type of district, like the park district, the industry district, plane district, etc. Maybe the game developer will add this in the future.

Thanks again for your work.

**Comment by hellas2000** [comment]
Hi TheDogKSP, 

I have just thought an idea. We can give each priority a balance value. For example, the lowest priority has a balance value 1000(this is a made up value, just for demonstrate), and the highest priority has a balance value 0, other priority has a balance value between 0~1000.

Then we can add the balance value to the distance value, the result value is then used to consider match-making.

This way priority and distance are both considered, and some special cases can work well too. 
For example, if a request has a high priority and a very long distance, the result value will still be high, and probably not match. 
If two requests have almost the same distance but different priority. The higher priority request have lower balance value, so the result value will be lower, and will probably be choosen.

The user can be given an option to set a balance factor value, use this value to determine balance values for each priority.
If the balance factor value be set properly, the problem should be solved.

Cheers.

**Comment by TheDogKSP** [comment]
@hellas2000 Havent seen you online in steam, but I have prepared a custom version for you.
Download here: https://github.com/TheDogKSP/MoreEffectiveTransfer/releases/tag/2.2.0-preview1-x

Then replace your steam workshop version with this.

It includes the changes for the upcoming 2.2.0 release, which is an improvedAI for garbage, police and fire vehicles (inspired by what SmartFireFighterAI does).

For you it also includes the change to prefer matching purely by closest distance as from demand side to possible supply.
I doubt you will see a lot of improvement in your save, but we'll see... :)

**Comment by hellas2000** [comment]
@TheDogKSP ,I have tried your new version for 4 hours on the map, still no luck, and devastating, here is a screenshot.
![1](https://user-images.githubusercontent.com/12899512/160639595-1a583481-26fc-4731-83f5-419257e4ea4f.jpg)

**Comment by TheDogKSP** [comment]
Thanks for testing that, I somewhat expected such results.
The urgent dead body pickups are not getting serviced as closer requests are occupying all the hearses.

The "more vanilla" algorithm with highest priority first (and only then modified by closest distance) just performs better.

Just to clarify: if you switch back to the workshop version of METM continued, then these death problems do not accumulate so much, is it?

**Comment by hellas2000** [comment]
> Just to clarify: if you switch back to the workshop version of METM continued, then these death problems do not accumulate so much, is it?

Yes, exactly, with this custom version, the death problem is much worse. Today I have tried it on another big map, the result is scary:(
